src/coreclr/jit/importercalls.cpp

```diff
@@ -699,7 +699,51 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
 
     if (sig->isAsyncCall())
     {
-        call->AsCall()->SetIsAsync();
+        AsyncCallInfo asyncInfo;
+
+        JITDUMP("Call is an async ");
+
+        if ((prefixFlags & PREFIX_IS_TASK_AWAIT) != 0)
+        {
+            JITDUMP("task await\n");
+
+            asyncInfo.ExecutionContextHandling = ExecutionContextHandling::SaveAndRestore;
+        }
+        else
+        {
+            JITDUMP("non-task await\n");
+            // Only expected non-task await to see in IL is one of the AsyncHelpers.AwaitAwaiter variants.
+            // These are awaits of custom awaitables, and they come with the behavior that the execution context
+            // is captured and restored on suspension/resumption.
+            // We could perhaps skip this for AwaitAwaiter (but not for UnsafeAwaitAwaiter) since it is expected
+            // that the safe INotifyCompletion will take care of flowing ExecutionContext.
+            asyncInfo.ExecutionContextHandling = ExecutionContextHandling::AsyncSaveAndRestore;
+        }
+
+        // For tailcalls the context does not need saving/restoring: it will be
+        // overwritten by the caller anyway.
+        //
+        // More specifically, if we can show that
+        // Thread.CurrentThread._executionContext is not accessed between the
+        // call and returning then we can omit save/restore of the execution
+        // context. We do not do that optimization yet.
+        if (tailCallFlags != 0)
```

Comment:
We still might decide not to actually tail call but I guess if the call is in tail position there's not going to be any code in this method after the call that will look at the execution context.